homeassistant/components/volkswagencarnet/__init__.py

```diff
@@ -0,0 +1,237 @@
+                    return False
+            else:
+                if not connection.update(request_data = False):
+                    _LOGGER.warning("Could not query update from volkswagen carnet")
+                    return False
+                else:
+                    _LOGGER.debug("Updating data from volkswagen carnet")
+
+                    for vehicle in connection.vehicles:
+                        if vehicle.vin not in data.vehicles:
+                            _LOGGER.info("Adding data for VIN: %s from carnet" % vehicle.vin.lower())
+                            discover_vehicle(vehicle)
+                        
+                        for entity in data.entities[vehicle.vin]:
+                            entity.schedule_update_ha_state()
+
+                        dispatcher_send(hass, SIGNAL_STATE_UPDATED, vehicle)
+            return True
+
+        finally:
+            track_point_in_utc_time(hass, update, utcnow() + interval)
+
+    _LOGGER.info("Starting volkswagencarnet component")
+    return update(utcnow())
+
+
+class VolkswagenData:
+    """Hold component state."""
+
+    def __init__(self, config):
+        """Initialize the component state."""
+        self.vehicles = set()
+        self.instruments = set()
+        self.entities = {}
+        self.config = config[DOMAIN]
+        self.names = self.config.get(CONF_NAME)
+
+    def instrument(self, vin, component, attr):
+        """Return corresponding instrument."""
+        return next((instrument
+                     for instrument in self.instruments
+                     if instrument.vehicle.vin == vin and
+                     instrument.component == component and
+                     instrument.attr == attr), None)
+
+    def vehicle_name(self, vehicle):
+        """Provide a friendly name for a vehicle."""
+        if (vehicle.vin and vehicle.vin.lower() in self.names):
+            return self.names[vehicle.vin.lower()]
+        elif vehicle.vin:
+            return vehicle.vin
+        else:
+            return ''
+
+
+class VolkswagenEntity(Entity):
+    """Base class for all Volkswagen entities."""
+
+    def __init__(self, data, vin, component, attribute):#, attribute):
+        """Initialize the entity."""
+        self.data = data
+        self.vin = vin
+        self.component = component
+        self.attribute = attribute
+
+        self.data.entities[self.vin].append(self)
+
+    @property
+    def instrument(self):
+        """Return corresponding instrument."""
+        return self.data.instrument(self.vin, self.component, self.attribute)
+
+    @property
+    def icon(self):
+        """Return the icon."""
+        if self.instrument.attr in ['battery_level', 'charging']:
+            return icon_for_battery_level(battery_level = self.instrument.state, charging = self.vehicle.charging)
+        else:
+            return self.instrument.icon
```

Comment:
Home Assistant uses a specific icon set. It seems weird to return the icon string directly from the library. It's fragile to couple the two APIs that hard.